### PR TITLE
Flatten allOf properties for OpenAI compatibility

### PR DESCRIPTION
**Pull request by pulphix**
## Summary

This PR adds an `allOf` flattener for JSON Schemas used by PydanticAI tools and structured outputs, and keeps the repo compliant with `contributing.md` (tests + pre-commit all green).

## Motivation

Many OpenAI(-compatible) providers, especially in strict structured-output mode, expect a flat object schema and either reject or mis-handle JSON Schema combinators like `allOf` / `oneOf` / `anyOf` in tool/function parameters. Without flattening, perfectly valid schemas generated by Pydantic can cause API errors or unstable model behavior.

### allOf flattener for object schemas

In `_json_schema.py`, we add logic to rewrite `allOf` between object schemas into a single `type: "object"` schema that is semantically equivalent in the supported cases:

- Merges `properties`
- Unions `required`
- Merges `patternProperties`
- Combines `additionalProperties` conservatively:
  - `false` only if all members have `additionalProperties: false`
  - otherwise, falls back to a more permissive setting

Additional behavior:

- The transformation is **recursive**: it also applies to nested objects and array items.
- If an `allOf` contains non-object members or obviously conflicting structures, the flattener **leaves it untouched** to avoid changing semantics.

### Provider integration

- `OpenAIJsonSchemaTransformer` now applies the flattener at the start of `transform`, so schemas passed to OpenAI (and OpenAI-compatible providers) no longer contain object-level `allOf` before strict-mode adjustments.
- `GoogleJsonSchemaTransformer` does the same before applying Gemini-specific schema tweaks (inlining `$defs`, removing unsupported keywords, etc.).

## Testing

- All existing tests pass.
- New tests were added/updated to cover:
  - Flattening of simple and nested object-level `allOf`
  - Preservation of schemas when `allOf` contains non-object members or conflicting structures
- `pre-commit` hooks run clean, keeping the repo compliant with `contributing.md`.

### CODE REVIEW

File: pydantic_ai_slim/pydantic_ai/profiles/google.py [hunk 1]
```diff
@@ -35,6 +35,8 @@ def __init__(self, schema: JsonSchema, *, strict: bool | None = None):
         super().__init__(schema, strict=strict, prefer_inlined_defs=True, simplify_nullable_unions=True)
 
     def transform(self, schema: JsonSchema) -> JsonSchema:
+        # Flatten object-only allOf to improve compatibility
+        schema = flatten_allof(schema)
```
Comment (DouweM) [inline pydantic_ai_slim/pydantic_ai/profiles/google.py]: Let's not touch the Google transformer as we have a lot of changes underway in https://github.com/pydantic/pydantic-ai/pull/3357.

File: pydantic_ai_slim/pydantic_ai/profiles/openai.py [hunk 2]
```diff
@@ -157,11 +157,18 @@ def walk(self) -> JsonSchema:
...
+                result.update(result['$defs'][root_key])
+            else:
+                # Fallback to original if transformed version not available (shouldn't happen in normal flow)
+                result.update(self.defs.get(root_key) or {})
 
         return result
 
     def transform(self, schema: JsonSchema) -> JsonSchema:  # noqa C901
+        # First, flatten object-only allOf constructs to avoid unsupported combinators in strict mode
+        schema = flatten_allof(schema)
```
Comment (DouweM) [inline pydantic_ai_slim/pydantic_ai/profiles/openai.py]: Could this be implemented on `JsonSchemaTransformer` with a keyword argument we can pass in `__init__`, similar to existing transformation flags?

File: tests/models/test_openai.py [hunk 3]
```diff
@@ -1915,6 +1915,36 @@ class MyModel(BaseModel):
...
+                'properties': {'foo': {'type': 'string'}},
+                'required': ['foo'],
+            },
+        },
+    }
+
+    transformer = OpenAIJsonSchemaTransformer(schema, strict=True)
+
+    # Mock super().walk() to return a result without $defs to test the fallback path
+    with patch.object(transformer.__class__.__bases__[0], 'walk', return_value={'$ref': '#/$defs/MyModel'}):
```
Comment (DouweM) [inline tests/models/test_openai.py]: This test is a bit unclear to me and `transformer.__class__.__bases__[0]` feels a bit brittle. What scenario are we testing exactly? Can't we get there without patching?

**Comment by DouweM** [comment]
> Many OpenAI(-compatible) providers, especially in strict structured-output mode, expect a flat object schema and either reject or mis-handle JSON Schema combinators like `allOf` / `oneOf` / `anyOf` in tool/function parameters. Without flattening, perfectly valid schemas generated by Pydantic can cause API errors or unstable model behavior.

@pulphix Can you please share examples of schemas that currently result in API errors, that this PR fixes? If this only applies to OpenAI-"compatible" APIs and not OpenAI itself, we'll need to make sure it only applies to those APIs.

File: tests/models/test_openai_like_providers.py
Comment (DouweM) [inline tests/models/test_openai_like_providers.py:1]: I don't think we need a separate test file here

File: tests/models/test_openai_like_providers.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+            },
+            {
+                'type': 'object',
+                'properties': {'bar': {'type': 'integer'}},
+                'required': ['bar'],
+            },
+        ],
+    }
+
+    transformer = transformer_factory(schema)
```
Comment (DouweM) [inline tests/models/test_openai_like_providers.py]: This seems unnecessary as we only test one transformer :)

File: tests/models/test_openai_like_providers.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    transformed = transformer.walk()
+
+    # allOf should have been flattened by the transformer
+    assert 'allOf' not in transformed
```
Comment (DouweM) [inline tests/models/test_openai_like_providers.py]: Please use `== snapshot()` as much as possible so we get to see the exact result.

File: tests/test_json_schema_flattener.py [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+                'type': 'object',
+                'properties': {'b': {'type': 'integer'}},
+                'patternProperties': {
+                    '^suffix_': {'type': 'number'},
+                },
+            },
+        ],
+    }
+
+    flattened = flatten_allof(copy.deepcopy(schema))
```
Comment (DouweM) [inline tests/test_json_schema_flattener.py]: I'd prefer to not use the function directly, only the transformer classes.

And as mentioned above: please use a single `snapshot()` assertion (using the inline-snapshot library) rather than many asserts.

**Comment by pulphix** [comment]
> > Many OpenAI(-compatible) providers, especially in strict structured-output mode, expect a flat object schema and either reject or mis-handle JSON Schema combinators like `allOf` / `oneOf` / `anyOf` in tool/function parameters. Without flattening, perfectly valid schemas generated by Pydantic can cause API errors or unstable model behavior.
> 
> @pulphix Can you please share examples of schemas that currently result in API errors, that this PR fixes? If this only applies to OpenAI-"compatible" APIs and not OpenAI itself, we'll need to make sure it only applies to those APIs.

I'm trying to integrate with notion MCP and getting the following error:
```
pydantic_ai.exceptions.ModelHTTPError: status_code: 400, model_name: gpt-5-nano-2025-08-07, body: {'message': "Invalid schema for function 'notion__notion-update-page': In context=('properties', 'data'), 'allOf' is not permitted.", 'type': 'invalid_request_error', 'param': 'tools[28].function.parameters', 'code': 'invalid_function_parameters'}
```

```json
{
  "type": "function",
  "function": {
    "name": "notion__notion-update-page",
    "description": "## Overview\nUpdate a Notion page's properties or content.\n## Properties\nNotion page properties are a JSON map of property names to SQLite values.\nFor pages in a database:\n- ALWAYS use the \"fetch\" tool first to get the data source schema and the\texact property names.\n- Provide a non-null value to update a property's value.\n- Omitted properties are left unchanged.\n\n**IMPORTANT**: Some property types require expanded formats:\n- Date properties: Split into \"date:{property}:start\", \"date:{property}:end\" (optional), and \"date:{property}:is_datetime\" (0 or 1)\n- Place properties: Split into \"place:{property}:name\", \"place:{property}:address\", \"place:{property}:latitude\", \"place:{property}:longitude\", and \"place:{property}:google_place_id\" (optional)\n- Number properties: Use JavaScript numbers (not strings)\n- Checkbox properties: Use \"__YES__\" for checked, \"__NO__\" for unchecked\n\n**Special property naming**: Properties named \"id\" or \"url\" (case insensitive) must be prefixed with \"userDefined:\" (e.g., \"userDefined:URL\", \"userDefined:id\")\nFor pages outside of a database:\n- The only allowed property is \"title\",\twhich is the title of the page in inline markdown format.\n\n## Content\nNotion page content is a string in Notion-flavored Markdown format. See the \"create-pages\" tool description for the full enhanced Markdown spec.\nBefore updating a page's content with this tool, use the \"fetch\" tool first to get the existing content to find out the Markdown snippets to use in the \"replace_content_range\" or \"insert_content_after\" commands.\n## Examples\n\t\t<example description=\"Update page properties\">\n\t\t{\n\t\t\t\"page_id\": \"f336d0bc-b841-465b-8045-024475c079dd\",\n\t\t\t\"command\": \"update_properties\",\n\t\t\t\"properties\": {\n\t\t\t\t\"title\": \"New Page Title\",\n\t\t\t\t\"status\": \"In Progress\",\n\t\t\t\t\"priority\": 5,\n\t\t\t\t\"checkbox\": \"__YES__\",\n\t\t\t\t\"date:deadline:start\": \"2024-12-25\",\n\t\t\t\t\"date:deadline:is_datetime\": 0,\n\t\t\t\t\"place:office:name\": \"HQ\",\n\t\t\t\t\"place:office:latitude\": 37.7749,\n\t\t\t\t\"place:office:longitude\": -122.4194\n\t\t\t}\n\t\t}\n\t\t</example>\n\t\t<example description=\"Replace the entire content of a page\">\n\t\t{\n\t\t\t\"page_id\": \"f336d0bc-b841-465b-8045-024475c079dd\",\n\t\t\t\"command\": \"replace_content\",\n\t\t\t\"new_str\": \"# New Section\nUpdated content goes here\"\n\t\t}\n\t\t</example>\n\t\t<example description=\"Replace specific content in a page\">\n\t\t{\n\t\t\t\"page_id\": \"f336d0bc-b841-465b-8045-024475c079dd\",\n\t\t\t\"command\": \"replace_content_range\",\n\t\t\t\"selection_with_ellipsis\": \"# Old Section...end of section\",\n\t\t\t\"new_str\": \"# New Section\nUpdated content goes here\"\n\t\t}\n\t\t</example>\n\t\t<example description=\"Insert content after specific text\">\n\t\t{\n\t\t\t\"page_id\": \"f336d0bc-b841-465b-8045-024475c079dd\",\n\t\t\t\"command\": \"insert_content_after\",\n\t\t\t\"selection_with_ellipsis\": \"## Previous section...\",\n\t\t\t\"new_str\": \"\n## New Section\nContent to insert goes here\"\n\t\t}\n\t\t</example>\n**Note**: For selection_with_ellipsis, provide only the first ~10 characters, an ellipsis, and the last ~10 characters. Ensure the selection is unique; use longer snippets if needed to avoid ambiguity.",
    "parameters": {
      "type": "object",
      "properties": {
        "data": {
          "allOf": [
            {
              "type": "object",
              "properties": {
                "page_id": {
                  "type": "string",
                  "description": "The ID of the page to update, with or without dashes."
                }
              },
              "required": [
                "page_id"
              ]
            },
            {
              "anyOf": [
                {
                  "type": "object",
                  "properties": {
                    "command": {
                      "type": "string",
                      "enum": [
                        "update_properties"
                      ]
                    },
                    "properties": {
                      "type": "object",
                      "additionalProperties": {
                        "type": [
                          "string",
                          "number",
                          "null"
                        ]
                      },
                      "description": "A JSON object that updates the page's properties. For pages in a database, use the SQLite schema definition shown in <database>. For pages outside of a database, the only allowed property is \"title\", which is the title of the page in inline markdown format. Use null to remove a property's value."
                    }
                  },
                  "required": [
                    "command",
                    "properties"
                  ],
                  "additionalProperties": false
                },
                {
                  "type": "object",
                  "properties": {
                    "command": {
                      "type": "string",
                      "enum": [
                        "replace_content"
                      ]
                    },
                    "new_str": {
                      "type": "string",
                      "description": "The new string to replace all content with."
                    }
                  },
                  "required": [
                    "command",
                    "new_str"
                  ],
                  "additionalProperties": false
                },
                {
                  "type": "object",
                  "properties": {
                    "command": {
                      "type": "string",
                      "enum": [
                        "replace_content_range"
                      ]
                    },
                    "selection_with_ellipsis": {
                      "type": "string",
                      "description": "Unique start and end snippet of the string to replace in the page content, including whitespace. DO NOT provide the entire string to replace. Instead, provide up to the first ~10 characters of the string to replace, an ellipsis, and then up to the last ~10 characters of the string to replace. Make sure you provide enough of the start and end snippet to uniquely identify the string to replace. For example, to replace an entire section, use \"old_start_and_end_snippet\":\"# Section heading...last paragraph.\""
                    },
                    "new_str": {
                      "type": "string",
                      "description": "The new string to replace the old string with."
                    }
                  },
                  "required": [
                    "command",
                    "selection_with_ellipsis",
                    "new_str"
                  ],
                  "additionalProperties": false
                },
                {
                  "type": "object",
                  "properties": {
                    "command": {
                      "type": "string",
                      "enum": [
                        "insert_content_after"
                      ]
                    },
                    "selection_with_ellipsis": {
                      "type": "string",
                      "description": "Unique start and end snippet of the string to match in the page content, including whitespace. DO NOT provide the entire string to match. Instead, provide up to the first ~10 characters of the string to match, an ellipsis, and then up to the last ~10 characters of the string to match. Make sure you provide enough of the start and end snippet to uniquely identify the string to match. For example, to match an entire section, use \"selection_with_ellipsis\":\"# Section heading...last paragraph.\""
                    },
                    "new_str": {
                      "type": "string",
                      "description": "The new content to insert."
                    }
                  },
                  "required": [
                    "command",
                    "selection_with_ellipsis",
                    "new_str"
                  ],
                  "additionalProperties": false
                }
              ]
            }
          ],
          "description": "The data required for updating a page"
        }
      },
      "required": [
        "data"
      ],
      "additionalProperties": false
    },
    "strict": true
  }
}
```

File: tests/test_json_schema_flattener.py (hunk 6 above)
Comment (pulphix) [reply tests/test_json_schema_flattener.py]: Fixed

File: tests/models/test_openai_like_providers.py (hunk 5 above)
Comment (pulphix) [reply tests/models/test_openai_like_providers.py]: Fixed

File: tests/models/test_openai_like_providers.py (hunk 4 above)
Comment (pulphix) [reply tests/models/test_openai_like_providers.py]: Fixed

File: tests/models/test_openai_like_providers.py
Comment (pulphix) [reply tests/models/test_openai_like_providers.py:1]: Fixed

File: tests/models/test_openai.py (hunk 3 above)
Comment (pulphix) [reply tests/models/test_openai.py]: Fixed

File: pydantic_ai_slim/pydantic_ai/profiles/openai.py (hunk 2 above)
Comment (pulphix) [reply pydantic_ai_slim/pydantic_ai/profiles/openai.py]: Fixed

File: pydantic_ai_slim/pydantic_ai/profiles/google.py (hunk 1 above)
Comment (pulphix) [reply pydantic_ai_slim/pydantic_ai/profiles/google.py]: Removed Change

File: pydantic_ai_slim/pydantic_ai/_json_schema.py [hunk 7]
```diff
@@ -4,12 +4,14 @@
...
 from copy import deepcopy
 from dataclasses import dataclass
-from typing import Any, Literal
+from typing import Any, Literal, cast
 
 from .exceptions import UserError
 
 JsonSchema = dict[str, Any]
 
+__all__ = ['JsonSchemaTransformer', 'InlineDefsJsonSchemaTransformer', 'flatten_allof']
```
Comment (DouweM) [inline pydantic_ai_slim/pydantic_ai/_json_schema.py]: Let's make `flatten_allof` private and not export it

File: pydantic_ai_slim/pydantic_ai/_json_schema.py [hunk 8]
```diff
@@ -188,3 +198,108 @@ def __init__(self, schema: JsonSchema, *, strict: bool | None = None):
...
+
+
+def _recurse_flatten_allof(schema: JsonSchema) -> JsonSchema:
+    s = deepcopy(schema)
+    s = _flatten_current_level(s)
+    s = _recurse_children(s)
+    return s
+
+
+def flatten_allof(schema: JsonSchema) -> JsonSchema:
```
Comment (DouweM) [inline pydantic_ai_slim/pydantic_ai/_json_schema.py]: We may not need this method at all if the JSON transformer can call `_recurse_flatten_allof` directly

File: tests/models/test_openai.py [hunk 9]
```diff
@@ -1915,6 +1915,111 @@ class MyModel(BaseModel):
...
+            },
+        },
+    }
+
+    transformer = OpenAIJsonSchemaTransformer(schema, strict=True)
+    result = transformer.walk()
+
+    # The transformer should resolve the $ref and use the transformed schema from $defs
+    # (not the original self.defs, which was the bug we fixed)
+    assert isinstance(result, dict)
```
Comment (DouweM) [inline tests/models/test_openai.py]: Please use `snapshot`, I want to see the entire result

File: tests/models/test_openai.py [hunk 10]
```diff
@@ -1915,6 +1915,111 @@ class MyModel(BaseModel):
...
+            },
+        },
+    }
+
+    transformer = OpenAIJsonSchemaTransformer(schema, strict=True)
+
+    # Simulate edge case: super().walk() returns $defs without root_key
+    # This shouldn't happen in normal flow, but we test the fallback path
+    with patch.object(
+        transformer.__class__.__bases__[0],
```
Comment (DouweM) [inline tests/models/test_openai.py]: I don't feel comfortable with this; if we can't come up with a real schema that gets us to that line, do we need it at all?

File: tests/test_json_schema_flattener.py [hunk 11]
```diff
@@ -0,0 +1,401 @@
...
+    flattened = transformer.walk()
+
+    assert flattened == snapshot(
+        {
+            'type': 'object',
+            'properties': {
+                'a': {'type': 'string'},
+                'b': {'type': 'integer'},
+            },
+            'additionalProperties': True,  # When any member has dict additionalProperties, result should be True
```
Comment (DouweM) [inline tests/test_json_schema_flattener.py]: Is that true? If I understand the `allOf` schema correctly, it would only accept an object with a `b` key and no other keys, or an empty object.

File: tests/test_json_schema_flattener.py [hunk 12]
```diff
@@ -0,0 +1,401 @@
...
+        'allOf': [
+            {
+                'type': 'object',
+                'properties': {'a': {'type': 'string'}},
+                'required': ['a'],
+            },
+            {
+                'type': 'object',
+                # No properties, required, or patternProperties
+                'additionalProperties': False,
```
Comment (DouweM) [inline tests/test_json_schema_flattener.py]: Wouldn't this imply that no properties are allowed at all?

File: tests/models/test_openai.py (hunk 10 above)
Comment (pulphix) [reply tests/models/test_openai.py]: @DouweM that was the old code before my change so I kept it as a fallback.
If you think is not anymore necessary we can remove the fallback.

File: tests/test_json_schema_flattener.py (hunk 12 above)
Comment (pulphix) [reply tests/test_json_schema_flattener.py]: @DouweM 
Correct, I think is not a good test, I'll change it with a more realistic one

File: tests/models/test_openai.py (hunk 10 above)
Comment (DouweM) [reply tests/models/test_openai.py]: @pulphix If I understand the new code correctly, if a ref was in `self.defs` previously, it's now guaranteed to be in `result['$defs']`, so the change should be fine without the fallback. Although we could keep the `or {}` that we had previously just to be sure.

We use `self.defs` in a few more places though, so if we now can't rely on that anymore, should we also update those other uses? Or actually set `self.defs` to the updated value after flattening? It could be worth verifying the interaction between the new flag and `prefer_inlined_defs` or `recursive_defs`.

Comment (pulphix) [reply tests/models/test_openai.py]: I was able to test the fallback setting to The OpenAITransformer prefer_inlined_defs as False, much cleaner test
When prefer_inlined_defs is set to False the code will use the original schema:
https://github.com/pydantic/pydantic-ai/blob/468e60a5cf73ea38aeab912d87b7ff7522f6bb80/pydantic_ai_slim/pydantic_ai/_json_schema.py#L48-L56

File: tests/test_json_schema_flattener.py (hunk 11 above)
Comment (pulphix) [reply tests/test_json_schema_flattener.py]: I updated all tests to reflect the correct allOf logic

File: tests/models/test_openai.py (hunk 9 above)
Comment (pulphix) [reply tests/models/test_openai.py]: Done

File: pydantic_ai_slim/pydantic_ai/_json_schema.py (hunk 8 above)
Comment (pulphix) [reply pydantic_ai_slim/pydantic_ai/_json_schema.py]: Done

File: pydantic_ai_slim/pydantic_ai/_json_schema.py (hunk 7 above)
Comment (pulphix) [reply pydantic_ai_slim/pydantic_ai/_json_schema.py]: Done

File: pydantic_ai_slim/pydantic_ai/_json_schema.py [hunk 13]
```diff
@@ -26,14 +28,18 @@ def __init__(
...
         simplify_nullable_unions: bool = False,  # TODO (v2): Remove this, no longer used
+        flatten_allof: bool = False,
     ):
         self.schema = schema
 
         self.strict = strict
-        self.is_strict_compatible = True  # Can be set to False by subclasses to set `strict` on `ToolDefinition` when set not set by user explicitly
+        # Can be set to False by subclasses to set `strict` on `ToolDefinition`
+        # when not set explicitly by the user.
+        self.is_strict_compatible = True
```
Comment (DouweM) [inline pydantic_ai_slim/pydantic_ai/_json_schema.py:44]: With this change, it's now ambiguous which line the comment belongs to :) Can you move it back please, even though the line is uncomfortable long? Or add a blank line between the 2 vars?

File: pydantic_ai_slim/pydantic_ai/_json_schema.py [hunk 14]
```diff
@@ -187,3 +197,274 @@ def __init__(self, schema: JsonSchema, *, strict: bool | None = None):
...
+        return {str(t) for t in cast(list[Any], schema_type)}
+    if isinstance(schema_type, str):
+        return {schema_type}
+    return None
+
+
+def _process_nested_schemas_without_allof(s: JsonSchema) -> JsonSchema:
+    """Process nested schemas recursively when there is no allOf at the current level."""
+    schema_type = s.get('type')
+    if schema_type == 'object':
```
Comment (DouweM) [inline pydantic_ai_slim/pydantic_ai/_json_schema.py:235]: We already have pretty similar code here:

https://github.com/pydantic/pydantic-ai/blob/1b576dd0c6de922aa1dac70d48128884ea2a1ae8/pydantic_ai_slim/pydantic_ai/_json_schema.py#L94-L101

Would it be an option to merge this new processing logic into that one, somehow? Or at least make them less seemingly-duplicative? I just don't love the idea of recursively walking the JSON schema twice with different implementations.

File: pydantic_ai_slim/pydantic_ai/profiles/openai.py [hunk 15]
```diff
@@ -157,7 +157,12 @@ def walk(self) -> JsonSchema:
         if self.root_ref is not None:
             result.pop('$ref', None)  # We replace references to the self.root_ref with just '#' in the transform method
             root_key = re.sub(r'^#/\$defs/', '', self.root_ref)
-            result.update(self.defs.get(root_key) or {})
+            # Use the transformed schema from $defs, not the original self.defs
+            if '$defs' in result and root_key in result['$defs']:
+                result.update(result['$defs'][root_key])
+            else:
+                # Fallback to original if transformed version not available (shouldn't happen in normal flow)
+                result.update(self.defs.get(root_key) or {})
```
Comment (DouweM) [inline pydantic_ai_slim/pydantic_ai/profiles/openai.py]: Please check the comment I had on this before; I don't think it's been addressed or at least I don't understand why we still have these 2 paths instead of standardizing in only using `self.defs`

**Review by DouweM — CHANGES_REQUESTED** [review]

File: pydantic_ai_slim/pydantic_ai/profiles/openai.py (hunk 15 above)
Comment (pulphix) [reply pydantic_ai_slim/pydantic_ai/profiles/openai.py]: Hi @DouweM, 7 months ago has been introduced this change:
https://github.com/pydantic/pydantic-ai/blob/5544e9fe4132c1b5b7eddc4f6a31e3d5c58ce4e7/pydantic_ai_slim/pydantic_ai/_json_schema.py#L54C1-L55C82
We should check the reason why has been introduced.

File: pydantic_ai_slim/pydantic_ai/_json_schema.py (hunk 14 above)
Comment (pulphix) [reply pydantic_ai_slim/pydantic_ai/_json_schema.py:235]: I refactored the code

File: pydantic_ai_slim/pydantic_ai/_json_schema.py (hunk 13 above)
Comment (pulphix) [reply pydantic_ai_slim/pydantic_ai/_json_schema.py:44]: Fixed

File: pydantic_ai_slim/pydantic_ai/_json_schema.py [hunk 16]
```diff
@@ -26,14 +29,16 @@ def __init__(
         strict: bool | None = None,
         prefer_inlined_defs: bool = False,
         simplify_nullable_unions: bool = False,  # TODO (v2): Remove this, no longer used
+        flatten_allof: bool = False,
     ):
         self.schema = schema
 
-        self.strict = strict
-        self.is_strict_compatible = True  # Can be set to False by subclasses to set `strict` on `ToolDefinition` when set not set by user explicitly
+        self.strict = strict  # Can be set to False by subclasses to set `strict` on `ToolDefinition `when not set explicitly by the user.
```
Comment (DouweM) [inline pydantic_ai_slim/pydantic_ai/_json_schema.py]: This actually belonged on the next line :D

File: pydantic_ai_slim/pydantic_ai/profiles/openai.py (hunk 15 above)
Comment (DouweM) [reply pydantic_ai_slim/pydantic_ai/profiles/openai.py]: @pulphix I don't understand, I see these lines as having been added by you in this PR:

Screenshot 2025-11-25 at 18 57 06

So I don't understand why this is necessary now if it wasn't before, and if it can't be simplified.

File: tests/models/test_openai.py [hunk 17]
```diff
@@ -1997,6 +1997,124 @@ class MyModel(BaseModel):
...
+            },
+        },
+    }
+
+    transformer = TestTransformer(schema, strict=True)
+    result = transformer.walk()
+
+    # When prefer_inlined_defs=True and model is not recursive, root_key won't be in result['$defs']
+    # So the fallback uses self.defs (original, untransformed schema)
+    # Note: The fallback uses the original schema, so it won't have additionalProperties: False
```
Comment (DouweM) [inline tests/models/test_openai.py:2070]: These 2 lines don't appear to be true anymore

File: pydantic_ai_slim/pydantic_ai/_json_schema.py [hunk 18]
```diff
@@ -197,3 +189,289 @@ def __init__(self, schema: JsonSchema, *, strict: bool | None = None):
...
+
+    properties, required, pattern_properties, additional_values, restricted_property_sets = _collect_base_schema_data(
+        result
+    )
+
+    # Then merge properties from all members
+    members_properties: list[dict[str, JsonSchema]] = [base_properties]
+    members_additional_props: list[Any] = [base_additional]
+
+    _collect_member_data(
```
Comment (DouweM) [inline pydantic_ai_slim/pydantic_ai/_json_schema.py:431]: It's not obvious that this modifies some of these arguments in place.

Can we please create a new helper `dataclass` that has the fields in question on it, and pass that around instead of separate vars? We could also have methods on that dataclass to do some of this work, which'll be easier to follow

File: tests/test_json_schema.py [hunk 19]
```diff
@@ -49,3 +49,25 @@ def transform(self, schema: dict[str, Any]) -> dict[str, Any]:
     # Should keep anyOf since it's not nullable
     assert 'anyOf' in result3
     assert len(result3['anyOf']) == 2
+
+
+def test_handle_object_preserves_non_mapping_properties_sections() -> None:
+    """Ensures properties/patternProperties that are present but not mappings are left untouched (cov 204-211, 217)."""
```
Comment (DouweM) [inline tests/test_json_schema.py:55]: These line numbers will be outdated/meaningless soon, so please remove them

**Comment by DouweM** [comment]
@pulphix Let me know if you're still interested in working on this and I'll reopen the PR!
